https://issues.redhat.com/browse/ACM-18037 (#7600)
open-cluster-management/rhacm-docs: governance/adv_template_process.adoc

    {{- end }}
----

[#hub-template-config-policies]

> == Using hub templates to define configuration policies

Reviewer: I think the title should be more like "Using hub templates without a hub Policy" - or maybe after re-wording the top paragraph, something better will be obvious.

> Use the `governance-standalone-hub-templating` addon in the `ManagedClusterAddOn` to define `ConfigurationPolicies` on your managed clusters. 

Reviewer: It's difficult to succinctly describe when users will want this. Maybe:

By default, using hub templates in `ConfigurationPolicy` and `OperatorPolicy` resources requires corresponding `Policy` resources defined on the hub cluster. When defining policies directly on managed clusters, for example via ArgoCD (OpenShift GitOps), you can enable the `governance-standalone-hub-templating` addon to resolve hub templates from the managed cluster.

To enable the `governance-standalone-hub-templating` addon, complete the following steps: 

. On your hub cluster, go to the managed cluster namespace. 
. Create a `ManagedClusterAddOn` resource with the name, `governance-standalone-hub-templating`, by using the following YAML sample: 

+
[source,yaml]
----
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: ManagedClusterAddOn
metadata:
  name: governance-standalone-hub-templating
  namespace: <cluster name>
  labels:
    cluster.open-cluster-management.io/backup: ''

> spec: {}

Reviewer: I don't believe this is necessary, but I realized this will make it look more consistent with the other addons.
```suggestion
spec:
  installNamespace: open-cluster-management-agent-addon
```

To allow access to `Configmaps` in your managed cluster namespace on your hub cluster, complete the following steps: 

. Add the `Role` resource by running the following command: 

+
[source,bash]
----
oc create role -n <cluster name> cm-reader --verb=get,list,watch --resource=configmaps
----

. Add the `Rolebinding` by running the following command: 

+
[source,bash]
----
oc create rolebinding -n <cluster name> cm-reader-binding --role=cm-reader --group=system:open-cluster-management:cluster:<cluster name>:addon:governance-standalone-hub-templating
----

. To ensure these resources on the hub cluster get backed up and restored, add the following label to each resource that you create: `cluster.open-cluster-management.io/backup`. 

> After you add these resources, they resolve the hub templates, and the state of the policy gets saved on a secret on the managed clusters. This secret prevents interruptions if the hub cluster becomes temporarily unavailable to the managed cluster.

Reviewer: ```suggestion
After you add these resources, the addon can resolve the hub templates, and the state of the policy gets saved on a secret on the managed clusters. This secret prevents interruptions if the hub cluster becomes temporarily unavailable to the managed cluster.
```

    {{- end }}
----

[#define-config-policies]

> == Defining configuration policies on managed clusters

Reviewer: I'm still hoping for a better section title here. The section is not totally about defining configuration policies on managed clusters, it is more specifically about how to resolve hub templates ***after*** defining those policies (and it works for both ConfigurationPolicy and OperatorPolicy).

Maybe I'm overthinking it - does the title need to have a verb? Or can it just be something like "The standalone hub templating addon"?

> By default, if you use the hub templates in the `ConfigurationPolicy` and `OperatorPolicy` resources, you must correspond the `Policy` resources defined on the hub cluster with these resources. 

Reviewer 2: _that are_ defined. Please be sure to read the standard on this that I think I liked recently.

Reviewer 2: "resources defined" can read the resources are defining the thing.

What we mean is resources **that are defined** , or **resources, which are** defined.

Reviewer 2: in general just check for that when writing

> You can define configuration policies directly on the managed clusters, for example with {gitops-short}. To resolve hub templates on the managed cluster in that situation, you can enable the `governance-standalone-hub-templating` addon. 

Reviewer 2: I don't quite understand the "for example" here. What is the example? It doesn't seem to match the subject. Comma after the `for example,` but the example itself is a bit confusing to me.

Reviewer: OpenShift GitOps is one way to define configuration policies directly on the managed clusters. Another way would be to `oc apply` a file while connected directly to the managed cluster. Additional ways include other CI tools like Flux, Jenkins, Ansible, ...

> You can define configuration policies directly on the managed clusters. For example, if you use the `ArgoCD` resource in {gitops-short}, you can enable the `governance-standalone-hub-templating` addon to resolve hub templates from the managed cluster. 

Reviewer: The `ArgoCD` ***resource*** configures some behaviors of the Openshift GitOps controllers. Perhaps confusingly, the upstream project itself is called ArgoCD. The ***project*** (or the RH product of it, Openshift Gitops) is what might be used to define policy resources directly on the managed cluster, not the ***resource***.

```suggestion
You can define configuration policies directly on the managed clusters, for example with {gitops-short}. To resolve hub templates on the managed cluster in that situation, you can enable the `governance-standalone-hub-templating` addon. 
```

> To enable the `governance-standalone-hub-templating` addon, complete the following steps: 

Reviewer 2: ```suggestion
To enable the `governance-standalone-hub-templating` add-on, complete the following steps: 
```

Reviewer 2: throughout the doc we used add-on

. On your hub cluster, go to the managed cluster namespace. 

> . Create a `ManagedClusterAddOn` resource with the name, `governance-standalone-hub-templating`, by using the following YAML sample: 

Reviewer 2: ```suggestion
. Create a `ManagedClusterAddOn` resource with the `governance-standalone-hub-templating` name by using the following YAML sample: 
```

Reviewer 2: Please change these each time to reduce the commas and improve conciseness. In all examples like this.

+
[source,yaml]
----
apiVersion: addon.open-cluster-management.io/v1alpha1
kind: ManagedClusterAddOn
metadata:
  name: governance-standalone-hub-templating
  namespace: <cluster name>
  labels:
    cluster.open-cluster-management.io/backup: ''
spec:
  installNamespace: open-cluster-management-agent-addon
----

> By default, the agent on the managed cluster only has access to the `ManagedCluster` resources on the hub cluster. You can use the `.ManagedClusterLabels` template variable in the hub cluster templates inside the `ConfigurationPolicies` that are deployed directly to the managed cluster. If you want the hub template to access other resources, such as: the `lookup` or `fromConfigMap` function calls, you must add those specific permissions to the addon group. You can add these permissions through resources, such as: `Roles`, `ClusterRoles`, `RoleBindings`, and `ClusterRoleBindings`. The name of the addon group depends on the name of your managed cluster, but it has the following standard form: `system:open-cluster-management:cluster:<cluster name>:addon:governance-standalone-hub-templating`.

Reviewer 2: Please avoid these long paragraphs in general. Line 196 can be multiple paragraphs. (This is feedback in general because it comes up a lot. We have to reduce these long paragraphs.)

Reviewer 2: **Such as** should not have a colon behind it.

Reviewer 2: another add-on in this paragraph when you break it out a bit.

To allow access to `Configmaps` in your managed cluster namespace on your hub cluster, complete the following steps: 

. Add the `Role` resource by running the following command: 

+
[source,bash]
----
oc create role -n <cluster name> cm-reader --verb=get,list,watch --resource=configmaps
----

. Add the `Rolebinding` by running the following command: 

+
[source,bash]
----
oc create rolebinding -n <cluster name> cm-reader-binding --role=cm-reader --group=system:open-cluster-management:cluster:<cluster name>:addon:governance-standalone-hub-templating
----

. To ensure these resources on the hub cluster get backed up and restored, add the following label to each resource that you create: `cluster.open-cluster-management.io/backup`. 

> After you add these resources, the addon can resolve the hub templates, and the state of the policy gets saved on a secret on the managed clusters. This secret prevents interruptions if the hub cluster becomes temporarily unavailable to the managed cluster.

Reviewer 2: ```suggestion
After you add these resources, the add-on can resolve the hub templates, and the state of the policy gets saved on a secret on the managed clusters. This secret prevents interruptions if the hub cluster becomes temporarily unavailable to the managed cluster.
```
